src/api-client.ts

```diff
@@ -0,0 +1,114 @@
+import * as httpm from "@actions/http-client";
+
+const userAgent = "GitHub multi-repository variant analysis";
+
+export function getApiClient() {
+  return new httpm.HttpClient(userAgent, [], {
+    allowRetries: true,
+  });
+}
+
+interface InProgressProperties {
+  status: "in_progress";
+}
+
+interface SuccessProperties {
+  status: "succeeded";
+  sourceLocationPrefix: string;
+  resultCount: number;
+  databaseSHA: string;
+}
+
+interface FailureProperties {
+  status: "failed";
+  failureMessage: string;
+}
+
+type updateVariantAnalysisProperties =
+  | InProgressProperties
+  | SuccessProperties
+  | FailureProperties;
+
+export async function setVariantAnalysisRepoInProgress(
+  variantAnalysisId: number,
+  repoId: number
+): Promise<void> {
+  await updateVariantAnalysisStatus(variantAnalysisId, repoId, {
+    status: "in_progress",
+  });
+}
+
+export async function setVariantAnalysisRepoSucceeded(
+  variantAnalysisId: number,
+  repoId: number,
+  sourceLocationPrefix: string,
+  resultCount: number,
+  databaseSHA: string
+): Promise<void> {
+  await updateVariantAnalysisStatus(variantAnalysisId, repoId, {
+    status: "succeeded",
+    sourceLocationPrefix,
+    resultCount,
+    databaseSHA,
+  });
+}
+
+export async function setVariantAnalysisFailed(
+  variantAnalysisId: number,
+  repoId: number,
+  failureMessage: string
+): Promise<void> {
+  await updateVariantAnalysisStatus(variantAnalysisId, repoId, {
+    status: "failed",
+    failureMessage,
+  });
+}
+
+async function updateVariantAnalysisStatus(
+  variantAnalysisId: number,
+  repoId: number,
+  data: updateVariantAnalysisProperties
+): Promise<void> {
+  const http = getApiClient();
+
+  const url = `/codeql/variant-analyses/${variantAnalysisId}/repositories/${repoId}`;
+  const response = await http.patch(url, JSON.stringify(data));
```

Comment:
Would it make sense to use `patchJson` instead? That should automatically set the `Content-Type` header (which might be necessary for the GitHub API, I'm not sure), parse the response as JSON and reject the promise when the request returns a non-200 status code (which will throw an error since you're awaiting the call).

Reply:
Maybe, though the actual implementation seems to be a bit unusual so we might still need some custom error handling. Overall I think doing it manually might be just as easy, both in terms of amount of code and readability.

See [`patchJson`](https://github.com/actions/toolkit/blob/bc4be505973a6a7344bfd71e1b32f77e1755310c/packages/http-client/src/index.ts#L299-L321) and [`_processResponse`](https://github.com/actions/toolkit/blob/bc4be505973a6a7344bfd71e1b32f77e1755310c/packages/http-client/src/index.ts#L696-L769) for the code that implements this.

The behaviour that I see as odd is that in the case of a [404 response](https://github.com/actions/toolkit/blob/bc4be505973a6a7344bfd71e1b32f77e1755310c/packages/http-client/src/index.ts#L710) it resolves the promise as successful, but with a null `respnose.result` field. This is a bit unhelpful for us because we don't expect to hit 404s and if we do then we want it to be treated as an error.

Potentially we could rewrite as something like this:
```typescript
const response = await http.patchJson(url, data);
if (response.statusCode !== 204) {
  throw new Error(
    `Error while setting variant analysis as "${data.status}". Status code: ${response.statusCode}`
  );
}
```

What do other people think?

Reply:
That's a bit strange. It sounds like it's not worth it to me, but no strong opinions.

Reply:
Thanks for the comments! Since there's no super strong feelings, I've left this as-is for now. Perhaps something we could do in a follow-up PR if that makes sense 😃